File: 프로그래머스/lv1/92334. 신고 결과 받기/신고 결과 받기.py
def solution(id_list, report, k):
    answer = [0] * len(id_list)    
    reports = {x : 0 for x in id_list}

    for r in set(report):
        reports[r.split()[1]] += 1

    for r in set(report):
        if reports[r.split()[1]] >= k:
            answer[id_list.index(r.split()[0])] += 1

    return answer

# def solution(id_list, report, k):
#     counts = {iid: 0 for iid in id_list}
#     reporters = {iid: set() for iid in id_list}
    
#     for arr in report:
#         ifrom, ito = arr.split()
#         if ifrom not in reporters[ito]:
#             counts[ito] += 1
#             reporters[ito].add(ifrom)
    
#     res = {iid: 0 for iid in id_list}
#     for iid, cnts in counts.items():
#         if cnts >= k:
#             for ifrom in reporters[iid]:
#                 res[ifrom] += 1
#     return list(res.values())

File: 프로그래머스/lv1/92334. 신고 결과 받기/신고 결과 받기.java
import java.util.*;

class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        int[] answer = new int[id_list.length];
        Arrays.fill(answer, 0);
        Map<String, Set<String>> userIdMap = new HashMap<>();
        Map<String, Set<String>> bandIdMap = new HashMap<>();
        for (String ids : report) {
            String[] split = ids.split(" ");
            if (userIdMap.containsKey(split[0])) {
                userIdMap.get(split[0]).add(split[1]);
            } else {
                Set<String> bidSet = new HashSet<>();
                bidSet.add(split[1]);
                userIdMap.put(split[0], bidSet);
            }
            if (bandIdMap.containsKey(split[1])) {
                bandIdMap.get(split[1]).add(split[0]);
            } else {
                Set<String> ridSet = new HashSet<>();
                ridSet.add(split[0]);
                bandIdMap.put(split[1], ridSet);
            }

        }
        for (int i = 0; i < id_list.length; i++) {
            if (userIdMap.containsKey(id_list[i])) {
                for (String bid : userIdMap.get(id_list[i])) {
                    if (k <= bandIdMap.get(bid).size()) {
                        answer[i] += 1;
                    }
                }
            }
        }
        return answer;
    }
}